app/jobs/autoscaling.py

```diff
@@ -198,9 +208,133 @@ def set_desired_capacity(asg, compute_desired_instances, can_scale=True):
+
+def instances_to_rescue(asg, num_instances):
+    '''
+    Rescue the instances that have been draining for the least amount of time.
+    That way we can probably terminate the non-rescued instances sooner.
+    '''
+    draining_instances = get_draining_servers(asg)
+    draining_instances_sorted = [key for key, value in sorted(draining_instances.items(), key=itemgetter(1), reverse=True)]
+    return draining_instances_sorted[:num_instances]
+
+
+def get_draining_servers(asg):
+    ''' returns a map of draining instance IDs to the time they started draining '''
+    cmd = "aws ec2 describe-tags --filters 'Name=tag-key,Values={draining_tag}' 'Name=resource-id,Values={instance_ids}'"
+    cmd = cmd.format(draining_tag=DRAINING_TAG, instance_ids=','.join(instances_in(asg)))
+    tag_dict = json.loads(aws_command(cmd))
+    instance_dict = { item['ResourceId']: dateutil.parser.parse(item['Value']) for item in tag_dict['Tags'] if item['Key'] == DRAINING_TAG }
+    return instance_dict
+
+
+def count_running_alignment_jobs(asg):
+    ''' returns a map of instance IDs to number of jobs running on the instance '''
+    instance_ids = instances_in(asg)
+    cmd = "aws ec2 describe-tags --filters 'Name=resource-id,Values={list_instances}'"
+    cmd = cmd.format(list_instances=','.join(instance_ids))
+    tag_dict = json.loads(aws_command(cmd))
+    count_dict = { id: 0 for id in instance_ids}
+    expired_jobs = []
+    for item in tag_dict['Tags']:
+        if item['Key'].startswith(JOB_TAG_PREFIX):
+            job_tag = item['Key']
+            instance_id = item['ResourceId']
+            unixtime = int(item['Value'])
+            if time.time() - unixtime < ALIGNMENT_JOB_EXPIRATION_SECONDS:
+                count_dict[instance_id] += 1
+            else:
+                expired_jobs.append(job_tag)
+    if expired_jobs:
+        print "The following job tags have expired and will be deleted: " + ", ".join(expired_jobs)
+        delete_tags_from_instances(instance_ids, expired_jobs)
+    return count_dict
+
+
+def delete_tags_from_instances(instance_ids, tag_keys):
+    if not instance_ids or not tag_keys:
+        return
+    cmd = "aws ec2 delete-tags --resources {list_instances} --tags {list_tags}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), list_tags=' '.join(['Key='+tag for tag in tag_keys]))
+    aws_command(cmd)
+   
+
+def remove_termination_protection(instance_ids, asg):
+    if not instance_ids:
+        return
+    cmd = "aws autoscaling set-instance-protection --instance-ids {list_instances} --auto-scaling-group-name {asg_name} --no-protected-from-scale-in"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), asg_name=asg['AutoScalingGroupName'])
+    aws_command(cmd)
+
+
+def add_termination_protection(instance_ids, asg):
+    if not instance_ids:
+        return
+    cmd = "aws autoscaling set-instance-protection --instance-ids {list_instances} --auto-scaling-group-name {asg_name} --protected-from-scale-in"
```

Comment:
Is it possible to make "protected from scale in" a default, so that the only transition executed by our code would be remove that protection when an instance is finished draining?

Reply:
Yes, that's exactly right -- I have a diff in idseq-infra for that: https://github.com/chanzuckerberg/idseq-infra/pull/148
This function `add_termination_protection` is actually for when we decide to `stop_draining` because in the meantime we decided to scale back up. In that case we just add the termination protection back, and avoid killing instances only to then launch new ones.

Reply:
I don't think it's safe to add termination protection back.   Once it has been removed, the instance is past resurrection.

Reply:
Agreed - I will remove that